pandas/core/arrays/period.py

```diff
@@ -121,18 +122,30 @@ def freq(self, value):
 
     _attributes = ["freq"]
 
-    def __new__(cls, values, freq=None, **kwargs):
+    def __new__(cls, values, freq=None, dtype=None, **kwargs):
+
+        if freq is not None:
+            # coerce freq to freq object, otherwise it can be coerced
+            # elementwise, which is slow
+            freq = Period._maybe_convert_freq(freq)
+
+        freq = dtl.validate_dtype_freq(dtype, freq)
+
         if is_period_dtype(values):
             # PeriodArray, PeriodIndex
-            if freq is not None and values.freq != freq:
-                raise IncompatibleFrequency(freq, values.freq)
-            freq = values.freq
+            freq = dtl.validate_dtype_freq(values.dtype, freq)
             values = values.asi8
 
         elif is_datetime64_dtype(values):
-            # TODO: what if it has tz?
             values = dt64arr_to_periodarr(values, freq)
 
+        elif is_object_dtype(values) or isinstance(values, (list, tuple)):
+            # e.g. array([Period(...), Period(...), NaT])
+            values = np.array(values)
```

Comment:
what if this is an int array? or is that prohibited? (except via _from_ordinals)

Reply:
Then it gets passed through simple_new unchanged.